### PR TITLE
Improve visibility into the reasons for compaction.

### PR DESCRIPTION
**Pull request by riversand963**
Add `compaction_reason` as part of event log for event `compaction started`.
Add counters for each `CompactionReason`.

### CODE REVIEW

File: db/compaction_job.cc [hunk 1]
```diff
@@ -62,6 +62,39 @@
...
+      return "FIFOTtl";
+    case CompactionReason::kManualCompaction:
+      return "ManualCompaction";
+    case CompactionReason::kFilesMarkedForCompaction:
+      return "FilesMarkedForCompaction";
+    case CompactionReason::kBottommostFiles:
+      return "BottommostFiles";
+    case CompactionReason::kTtl:
+      return "Ttl";
+    default:
```
Comment (ajkr) [inline db/compaction_job.cc]: will compile succeed if you remove the "default" (you could make `kNumOfReasons` an `assert(false)` case)? It'll allow people who add `CompactionReason` to easily find they need to update this function.

On that note, I think we missed `kExternalSstIngestion` here

File: db/obsolete_files_test.cc [hunk 2]
```diff
@@ -53,7 +53,7 @@ class ObsoleteFilesTest : public testing::Test {
     options_.max_bytes_for_level_base = 1024*1024*1000;
     options_.WAL_ttl_seconds = 300; // Used to test log files
     options_.WAL_size_limit_MB = 1024; // Used to test log files
-    dbname_ = test::TmpDir() + "/double_deletefile_test";
+    dbname_ = test::TmpDir() + "/obsolete_files_test";
```
Comment (ajkr) [inline db/obsolete_files_test.cc]: good find

File: db/internal_stats.h [hunk 3]
```diff
@@ -235,6 +281,19 @@ class InternalStats {
       this->num_input_records -= c.num_input_records;
       this->num_dropped_records -= c.num_dropped_records;
       this->count -= c.count;
+      int numOfReasons = static_cast<int>(CompactionReason::kNumOfReasons);
```
Comment (ajkr) [inline db/internal_stats.h]: naming convention is lower-case underscore-separated for local variables (we mostly try to follow google style guide).

File: db/compaction_job.cc (hunk 1 above)
Comment (riversand963) [reply db/compaction_job.cc]: Good catch on this!
I think we can add the `case CompationReason::kNumOfReasons` above `default:` as follows:
```
case CompactionReason::kNumOfReasons:
// fall through
default:
  assert(false);
```

File: db/internal_stats.h (hunk 3 above)
Comment (riversand963) [reply db/internal_stats.h]: Done.

**Comment by riversand963** [comment]
I've addressed the current comments, and start working on adding unit tests for the stats. Will push a new version later. Thanks @ajkr

File: db/compaction_job.cc (hunk 1 above)
Comment (ajkr) [reply db/compaction_job.cc]: oh sure, that's good

File: db/internal_stats.cc [hunk 4]
```diff
@@ -471,6 +471,28 @@ bool InternalStats::GetIntPropertyOutOfMutex(
...
 
+/*
+ * This function performs sanity check on the compaction stats. This function
+ * should be used in tests only.
+ *
+ * This function verifies that the number of compactions caused by different
+ * reasons sum up to equal the total number of compactions.
+ * @return true if sanity check passes, false otherwise
+*/
+bool InternalStats::TEST_CompactionStatsSanityCheck() const {
```
Comment (ajkr) [inline db/internal_stats.cc]: is it possible to move this logic to the test file? I believe we usually use "TEST_*" functions to expose private members for convenience, not so much for the actual test logic.

**Pull request closed without merging**